go/vt/mysqlctl/mysqld.go

```diff
@@ -107,7 +112,52 @@ func NewMysqld(dbcfgs *dbconfigs.DBConfigs) *Mysqld {
 	result.appPool = dbconnpool.NewConnectionPool("AppConnPool", *appPoolSize, *appIdleTimeout, *poolDynamicHostnameResolution)
 	result.appPool.Open(dbcfgs.AppWithDB(), appMysqlStats)
 
-	return result
+	version, _ := getVersionString()
```

Comment:
I don't think we can make the version string compulsory. In the "remote mysqld" case, this version check may fail since there's no need for `mysqld` to be available locally on the system running vttablet. Even though we (hopefully) don't use the methods of the `Mysqld` type that depend on version info in the remote case, we still need to create and use a `Mysqld` object to call the other methods on it.

My suggestion is to make `mysqld.capabilities` a pointer and leave it `nil` if `getVersionString()` fails. This ensures that in the remote mysqld case, we can still create a `Mysqld` object, but if we mistakenly try to call any methods on it that depend on `capabilities`, we will panic rather than risk undefined behavior.

If `getVersionString()` succeeds, but `parseVersionString()` fails, we can bail out with an error here as you already do.

Reply:
I am not sure there is a "remote mysqld" case? As I understand `mysqlctl` is only used locally, and has some preexisting dependencies on it:

```
RunMySQLUpgrade(): calls binaryPath and executes mysql_upgrade locally
startNoWait(): executes `mysqld` / `mysqld_safe` (called by Start())
```

To paraphrase: I see `mysqlctl` as a wrapper around MySQL Server functionality that needs to be executed locally (initializing server, upgrading*, installing config files*).

(* In newer versions these can happen remotely, but we need to support 5.6+)

Reply:
It's true that the `mysqlctl` binary (in `go/cmd/mysqlctl`) is only used when running `mysqld` locally. However, parts of the `go/vt/mysqlctl` *package* are still used by `go/cmd/vttablet` even when not using `go/cmd/mysqlctl`. Creating a `mysqlctl.Mysqld` instance via `mysqlctl.NewMysqld()` is one of the parts that `go/cmd/vttablet` still relies on, even in the remote mysqld case.

> To paraphrase: I see mysqlctl as a wrapper around MySQL Server functionality that needs to be executed locally

It probably started that way, but the abstraction leaked over time, and now that's not the only thing the `go/vt/mysqlctl` package is needed for. Now there are some parts of `go/vt/mysqlctl` that are dark in the remote mysqld case, but other parts that are live. There is no easy way to tell from reading the code which parts are in which category.

Reply:
OK I understand better. Thanks for clarifying!

I took the path of reincorporating fallback logic based on `MySQL_FLAVOR`. The motivation was that it looks like these other packages do things related to the configuration, and this PR now automagically includes certain config files based on versions.

Since it gets a bit fancy, and `mysqld --version` can differ, I've also added tests.

There is a bit of an annoying case, where the `MYSQL_FLAVOR` of MySQL56 is treated as MySQL 5.7. There never was a `MYSQL_FLAVOR` of MySQL57, and a lot of the value of capabilities is lost if the values of `flavor` and `version` could loosely mean either.

This incompatibility means that MySQL 5.6 will not bootstrap when the `mysqld` binary can not be located. I believe this is a fairly minor case though.